### ISSUE TITLE
Exception "tensor does not have a device" after casting with to

### ISSUE THREAD

**Issue by nlsmrg** [pytorch/pytorch#34138]
## 🐛 Bug
`torch::nn::Linear` without bias throws an error when `to` is called. 

## To Reproduce
```c++
#include<torch/torch.h>
int main(int argc, char* argv[])
{
  torch::Tensor x = torch::ones({100, 4});
  auto lopt       = torch::nn::LinearOptions(4, 4).bias(false);
  auto lin        = torch::nn::Linear(lopt);
  // works
  auto y          = lin->forward(x);
  // error
  lin->to(torch::kFloat64);
  x = torch::ones({100, 4}, torch::kFloat64);
  y = lin->forward(x);
}
```
This is the output, but this MWE is only executed as a test in a larger project
so the output might differ.
```
 terminate called after throwing an instance of 'c10::Error'
  what():  tensor does not have a device (device at /home/nls/gasnew/pytorch/torch/include/c10/core/TensorImpl.h:461)
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6a (0x7ffff220035a in /home/nls/gasnew/pytorch/torch/lib/libc10.so)
frame #1: c10::TensorImpl::device() const + 0xf8 (0x7ffff7f0e288 in /home/nls/gasnew/gasPyTorch/build/gt/src/libgt.so)
frame #2: at::Tensor::device() const + 0x20 (0x7ffff7f117d4 in /home/nls/gasnew/gasPyTorch/build/gt/src/libgt.so)
frame #3: at::Tensor::options() const + 0x71 (0x7ffff7f5fb41 in /home/nls/gasnew/gasPyTorch/build/gt/src/libgt.so)
frame #4: at::native::to(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) + 0x6c (0x7ffff368fabc in /home/nls/gasnew/pytorch/torch/lib/libtorch.so)
frame #5: <unknown function> + 0x1714a6b (0x7ffff392aa6b in /home/nls/gasnew/pytorch/torch/lib/libtorch.so)
frame #6: <unknown function> + 0x32572d0 (0x7ffff546d2d0 in /home/nls/gasnew/pytorch/torch/lib/libtorch.so)
frame #7: <unknown function> + 0x173c003 (0x7ffff3952003 in /home/nls/gasnew/pytorch/torch/lib/libtorch.so)
frame #8: at::Tensor c10::KernelFunction::callUnboxedOnly<at::Tensor, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> >(at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) const + 0x141 (0x55555559bfd3 in /home/nls/gasnew/gasPyTorch/build/test/testgt)
frame #9: at::Tensor c10::Dispatcher::callUnboxedOnly<at::Tensor, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat> >(c10::OperatorHandle const&, at::Tensor const&, c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) const + 0x16f (0x5555555968c9 in /home/nls/gasnew/gasPyTorch/build/test/testgt)
frame #10: at::Tensor::to(c10::ScalarType, bool, bool, c10::optional<c10::MemoryFormat>) const + 0x1de (0x55555558f6dc in /home/nls/gasnew/gasPyTorch/build/test/testgt)
frame #11: void torch::nn::Module::to_impl<c10::ScalarType&, bool&>(c10::ScalarType&, bool&) + 0x149 (0x7ffff59d5f19 in /home/nls/gasnew/pytorch/torch/lib/libtorch.so)
frame #12: torch::nn::Module::to(c10::ScalarType, bool) + 0x1c (0x7ffff59d2bac in /home/nls/gasnew/pytorch/torch/lib/libtorch.so)
```
## Environment

 - PyTorch Version: 1.4 from source
 - OS: Linux

cc @yf225

**Comment by yf225** [comment]
This problem should have been fixed by https://github.com/pytorch/pytorch/pull/30523 and https://github.com/pytorch/pytorch/pull/30626, and the fixes should be in 1.4 release. I wasn't able to reproduce this error when using libtorch 1.4 downloaded from https://pytorch.org/ either. Do you mind giving the downloaded libtorch 1.4 a try?

**Comment by nlsmrg** [comment]
I applied some patches in order to use gcc-9.1.
Sorry didn't think about that, when I opened the issue.
I've now switched to the current nightly and the error is gone
and I can use gcc-9.
